### PR TITLE
feat: support bigquery

### PR DESCRIPTION
**Pull request by wd0517**
This PR supports replicating data from TiDB to [BigQuery](https://cloud.google.com/bigquery) by following [BigQuery's CDC Migration Guide](https://cloud.google.com/bigquery/docs/migration/database-replication-to-bigquery-using-change-data-capture).

### CODE REVIEW

File: pkg/bigquerysql/connector.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	tableExists, err := checkTableExists(ctx, bc.bqClient, bc.datasetID, bc.tableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if tableExists {
+		err = deleteTable(ctx, bc.bqClient, bc.datasetID, bc.tableID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/connector.go:103]: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#create_table_statement

Maybe we can use `create or replace` ?

File: pkg/bigquerysql/operation.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		return errors.Trace(err)
+	}
+	tableRef := client.Dataset(datasetID).Table(tableID)
+	if err := tableRef.Create(context.Background(), tableMetadata); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func createExternalTable(ctx context.Context, client *bigquery.Client, datasetID, externalTableID string, columns []cloudstorage.TableCol, storageURI *url.URL, filePath string) error {
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/operation.go:80]: seems defined but not used?

File: pkg/bigquerysql/types.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+func GetBigQueryColumnTypeString(column cloudstorage.TableCol) (string, error) {
+	colType, err := GetBigQueryColumnType(column)
+	if err != nil {
+		return "", err
+	}
+
+	switch colType {
+	case bigquery.FloatFieldType:
+		return "FLOAT64", nil
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/types.go:61]: why?

Comment (wd0517) [reply pkg/bigquerysql/types.go:61]: The poor design of BigQuery is the reason for this. In SQL, it only accepts `FLOAT64`, but in the [API](https://github.com/googleapis/google-cloud-go/blob/main/bigquery/schema.go#L253-L254), it accepts `FLOAT`.

image

File: pkg/bigquerysql/operation.go (hunk 2 above)
Comment (wd0517) [reply pkg/bigquerysql/operation.go]: removed

File: pkg/bigquerysql/types.go (hunk 3 above)
Comment (Lloyd-Pottiger) [reply pkg/bigquerysql/types.go:61]: okk...... add some comments?

File: pkg/bigquerysql/connector.go (hunk 1 above)
Comment (wd0517) [reply pkg/bigquerysql/connector.go]: Good idea!

BigQuery provides two methods for interacting with the database:
- [API](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#methods), which does not support create or replace, and have some bad design such as the use of `FLOAT`.
- SQL, as you mentioned.

I have now refactored all the DDL operations to use the `SQL` approach, which provides more comprehensive functionality. Please review again.

File: pkg/bigquerysql/connector.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(ddls) == 0 {
+		log.Info("No need to execute this DDL in BigQuery", zap.String("ddl", tableDef.Query))
+		return nil
+	}
+	// One DDL may be rewritten to multiple DDLs
+	ctx := context.Background()
+	for _, ddl := range ddls {
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/connector.go:67]: maybe add a context variable to struct?

File: pkg/bigquerysql/connector.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	createTableSQL, err := GenCreateSchema(tableColumns, pKColumns, bc.datasetID, bc.tableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err = runQuery(ctx, bc.bqClient, createTableSQL); err != nil {
+		return errors.Annotate(err, "Failed to create table")
+	}
+	return nil
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/connector.go:101]: print a info log here?

File: pkg/bigquerysql/connector.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+func (bc *BigQueryConnector) LoadSnapshot(targetTable, filePrefix string, onSnapshotLoadProgress func(loadedRows int64)) error {
+	ctx := context.Background()
+	gcsFilePath := fmt.Sprintf("%s/%s*.csv", bc.storageURL, filePrefix)
+	// FIXME: if source table is empty, bigquery will fail to load (file not found)
+	err := loadGCSFileToBigQuery(ctx, bc.bqClient, bc.datasetID, bc.tableID, gcsFilePath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/connector.go:112]: ditto

File: pkg/bigquerysql/ddl.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	return ddls, nil
+}
+
+func getDefaultString(val interface{}) string {
+	_, err := strconv.ParseFloat(fmt.Sprintf("%v", val), 64)
+	if err != nil {
+		return fmt.Sprintf("'%v'", val) // FIXME: escape
+	}
+	return fmt.Sprintf("%v", val)
+}
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/ddl.go:111]: Can bigquery handle `alter table t add column a int default "1"`? If yes, this function is not needed.

File: pkg/bigquerysql/sql.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/pkg/sink/cloudstorage"
+)
+
+var (
+	CDC_FLAG_COLUMN_NAME       = "tidb2dw_flag"
+	CDC_TABLENAME_COLUMN_NAME  = "tidb2dw_tablename"
+	CDC_SCHEMANAME_COLUMN_NAME = "tidb2dw_schemaname"
+	CDC_TIMESTAMP_COLUMN_NAME  = "tidb2dw_timestamp"
```
Comment (Lloyd-Pottiger) [inline pkg/bigquerysql/sql.go:15]: image

rename to `CDC_COMMIT_TS_COLUMN_NAME` ?

File: pkg/bigquerysql/connector.go (hunk 1 above)
Comment (Lloyd-Pottiger) [reply pkg/bigquerysql/connector.go]: 👍

File: pkg/bigquerysql/ddl.go (hunk 7 above)
Comment (wd0517) [reply pkg/bigquerysql/ddl.go:119]: When setting the default value of a string column to 'a', bigquery can not handle `alter table t alter column str_coulmn set default a`, so this function is still necessary.

**Review by Lloyd-Pottiger — APPROVED** [review]
Good job!

**Review by Lloyd-Pottiger — COMMENTED** [review]
Maybe update the README.md?

**Comment by wd0517** [comment]
> Maybe update the README.md?

The readme has been updated.

**Pull request merged**